### DISCUSSION TITLE
How to create a public store based on the original but without the setState method?

### DISCUSSION THREAD

**Discussion by budarin** [pmndrs/zustand#2056] [Q&A]
I'm trying to create a store with properties and methods exclusively with which you can modify the state
To do this, I need to create a facade function in which there will be no static setState method.

I'm trying to do it with this code:
```ts
const store = create<AppState & Actions>()(() => ({ ... }));

const useAppStore = Object.assign(
  (...args: Parameters<StoreType>): ReturnType<StoreType> => {
    return store(...args);
  },
  store
);

delete useAppStore.setState;
```

but the typescript writes an error:
```
The signature '(selector: (state: AppState & Actions) => unknown, equalityFn: (a: unknown, b: unknown) => boolean): unknown' of 'store' is deprecated.ts(6387)
react.d.ts(20, 8): The declaration was marked as deprecated here.
```

![image](https://github.com/pmndrs/zustand/assets/8055157/5feb58bc-57f8-451c-8099-c9ca6ec0f05c)

What is the reason?
How to fix it?

**Comment by dai-shi** [comment]
If you need to use `equalityFn`, try using `createWithEqualityFn` from 'zustand/traditional'.

Otherwise, what's your StoreType? Can you create a reproduction with https://www.typescriptlang.org/play ?

**Reply by budarin** [reply]
I don't need equalityfm!

I just don't understand why typescript swears at such a general syntax and it's not clear what to do in such a situation

**Reply by dai-shi** [reply]
what's your StoreType? Can you create a reproduction with https://www.typescriptlang.org/play ?

**Reply by budarin** [reply]
something like that

```ts
const store = create<AppState & AppActions>()(() => ({
  entity1: {...},
  addEntity1: (...) => { ...},
  createEntity1: (...) => { ...},
  updateEntity2: (...) => { ...},
  deleteEntity3: (...) => { ...},
  
  entity2: {...},
  addEntity1: (...)=> { ...},
  
  ...  
}))
```

**Reply by budarin** [reply]
the structure of the store absolutely does not matter - here the question is in typescript - it considers such syntax `rawStore(...args)` as deprecated

```ts
export const useAppStore = Object.assign((...args: Parameters<StoreType>): ReturnType<StoreType> => {
    return rawStore(...args);
}, rawStore);
```

**Reply by dbritto-dev** [reply]
@budarin what you are doing doesn't make sense to me if you are trying to create an store you should use `create` (static) or `createStore` (dynamic)? So far I see seems you are trying to create a store dynamically so you should use `createStore`

**Reply by budarin** [reply]
what is it?
```ts
const store = create<AppState & AppActions>()(() => ({ ... }))
```
is it not `create`?
Please, try to read the issue

**Reply by dbritto-dev** [reply]
@budarin object.assign + `create` if what you are using, just use `createStore` without exposing `setState()`

**Reply by budarin** [reply]
What do you mean by `createStore`?
I have'n found such API :(

**Reply by budarin** [reply]
setState - it's static props of the store
How do you suggest not to expose it?

**Reply by budarin** [reply]
I have no idea what you mean - can you show the code please

**Comment by budarin** [comment]
@dai-shi 

Please, create an API section in the documentation

**Reply by dai-shi** [reply]
Docs are maintained by community members. Would you like to work on it? I can review/answer if there's anything ambiguous. 

---
Found this comment: https://github.com/pmndrs/zustand/discussions/2056#discussioncomment-7027109

> What do you mean by `createStore`? I have'n found such API :(

createStore is described in two sections:
- https://github.com/pmndrs/zustand#using-zustand-without-react
- https://github.com/pmndrs/zustand#react-context
